interlok-apache-http5/src/main/java/com/adaptris/core/http/apache5/ApacheHttpProducer.java

```diff
@@ -71,98 +68,63 @@
     return ObjectUtils.defaultIfNull(getResponseHandlerFactory(), DEFAULT_HANDLER);
   }
 
   @Override
-  public synchronized void close() {
-    Closer.closeQuietly(httpClient);
-    httpClient = null;
-    super.close();
-  }
-
-  @Override
-  protected AdaptrisMessage doRequest(AdaptrisMessage msg, String uri, long timeout)
-      throws ProduceException {
-    AdaptrisMessage reply = defaultIfNull(getMessageFactory()).newMessage();
-
-    try (HttpAuthenticator auth = authenticator()) {
+  protected AdaptrisMessage doRequest(AdaptrisMessage msg, String uri, long timeout) throws ProduceException {
+    try (HttpAuthenticator auth = authenticator())
+    {
       HttpUriRequestBase httpOperation = getMethod(msg).create(uri);
       auth.setup(uri, msg, new ApacheResourceTargetMatcher(httpOperation.getUri()));
       log.trace("Attempting [{}] against [{}]", httpOperation.getMethod(), httpOperation.getUri());
-      httpClient = createClient(timeout, httpOperation);
-      if (auth instanceof ApacheRequestAuthenticator) {
+      HttpClient httpClient = getClient(httpOperation, getClientConfig(), timeout);
+      if (auth instanceof ApacheRequestAuthenticator)
+      {
         ((ApacheRequestAuthenticator) auth).configure(httpOperation);
       }
-      addData(msg, getRequestHeaderProvider().addHeaders(msg, httpOperation));
-      reply = httpClient.execute(httpOperation, responseHandlerFactory().createResponseHandler(this));
-      preserveRequestPayload(msg, reply);
-
-      httpOperation.getConfig();
+      getRequestHeaderProvider().addHeaders(msg, httpOperation).setEntity(new AdaptrisMessageEntity(msg, getContentTypeProvider()));
 
-    } catch (Exception e) {
+      AdaptrisMessage reply = httpClient.execute(httpOperation, responseHandlerFactory().createResponseHandler(this));
+      preserveRequestPayload(msg, reply);
+      return reply;
+    }
+    catch (Exception e)
+    {
       throw ExceptionHelper.wrapProduceException(e);
     }
-    return reply;
   }
 
-  private synchronized CloseableHttpClient createClient(long timeout, HttpUriRequestBase httpOperation) throws Exception {
-    if (httpClient == null) {
-      HttpClientBuilder builder = customise(
-              HttpClientBuilderConfigurator.defaultIfNull(getClientConfig())
-                      .configure(HttpClients.custom(), timeout));
-      httpClient = builder.build();
-    } else {
-      if (timeout <= 0) {
-        // no timeout, or use default
-        return httpClient;
-      }
+  private synchronized CloseableHttpClient getClient(HttpUriRequestBase httpOperation, HttpClientBuilderConfigurator clientConfig, long timeout) throws Exception
+  {
+    if (httpClient == null || clientConfig != null)
+    {
+      /*
+       * Either create an HTTP client instance or reconfigure with clientConfig.
+       */
+      httpClient = HttpClientBuilderConfigurator.defaultIfNull(clientConfig).configure(HttpClients.custom(), timeout).build();
```

Comment:
The producer will use an existing `HttpClient` if there is one, unless the user has set the advanced `clientConfig`. What I was aiming for was if the user never set `clientConfig` then every time the producer is called then a single instance of `HttpClient` is used (as is recommended by the Apache documentation). However, if the user does set `clientConfig` then they get a new `HttpCilent` instance configured as they wish but at the expense of actually creating a new instance. If I've missed the mark, then the `static HttpClient` can be removed and it'll just be a local variable.

Reply:
I get your point but the problem here is that if we have two workflows with each an http producer but they behave differently (different clientConfig), you may have some conflict because they will share the same httpClient.
I think you approach to not create a new http client is a good idea but maybe a class variable instead of a static variable.
I think that's still better than in the interlok-apache-http project (https://github.com/adaptris/interlok-apache-http/blob/develop/interlok-apache-http/src/main/java/com/adaptris/core/http/apache/ApacheHttpProducer.java) which seems to create the client every time but at least we will avoid any unwanted behaviour?

Reply:
> class variable instead of a static variable

This could be the way to go; I'll give it a try.